contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+  APIVersion string
+  // Object is the CRD object to be converted.
+  Object runtime.RawExtension
+}
+
+type ConversionResponse struct {
+  // UID is an identifier for the individual request/response.
+  // This should be copied over from the corresponding ConversionRequest.
+  UID types.UID
+  // ConvertedObject is the converted version of request.Object.
+  ConvertedObject runtime.RawExtension
+}
+```
+
+If the conversion is failed, the webhook should fail the HTTP request with a proper error code and message that will be used to create a status error for the original API caller.
+
+
+## Monitorability
+
+There should be prometheus variables to show:
+
+* CRD conversion latency
+    * Overall
+    * By webhook name
+    * By request (sum of all conversions in a request)
+    * By CRD
+* Conversion Failures count
+    * Overall
+    * By webhook name
+    * By CRD
+* Timeout failures count
+    * Overall
+    * By webhook name
+    * By CRD
+
+Adding a webhook dynamically adds a key to a map-valued prometheus metric. Webhook host process authors should consider how to make their webhook host monitorable: while eventually we hope to offer a set of best practices around this, for the initial release we won’t have requirements here.
+
+
+## Error Messages
+
+When a conversion webhook fails, e.g. for the GET operation, then the error message from the apiserver to its client should reflect that conversion failed and include additional information to help debug the problem. The error message and HTTP error code returned by the webhook should be included in the error message API server returns to the user.  For example:
+
+```bash
+$ kubectl get mykind somename
+error on server: conversion from stored version v1 to requested version v2 for somename: "408 request timeout" while calling service "mywebhookhost.somens.cluster.local:443"
+```
+
+
+For operations that need more than one conversion (e.g. LIST), no partial result will be returned. Instead the whole operation will fail the same way with detailed error messages. To help debugging these kind of operations, the UID of the first failing conversion will also be included in the error message. 
+
+
+## Caching
+
+No new caching is planned as part of this work, but the API Server may in the future cache webhook POST responses.
+
+Most API operations are reads.  The most common kind of read is a watch.  All watched objects are cached in memory.  Each GVK is cached, regardless of whether some GVKs share storage.  Thus, watches do not not cause conversion.  So, conversion webhooks will not add overhead to the watch path.  Watch cache is per api server and eventually consistent.
+
+Non-watch reads are also cached (if requested resourceVersion is 0 which is true for generated informers by default, but not for calls like `kubectl get ...`, namespace cleanup, etc). The cached objects are converted and per version (TODO: fact check). So, conversion webhooks will not add overhead here too.
```

Comment:
original comment by @liggitt  need to be addressed?

> it does still add overhead, but I think the watch cache makes it scale with the number of versions * number of apiservers, not by the number of overall watchers

Reply:
Here is the code: https://github.com/kubernetes/kubernetes/blob/dbb448bbdcb9e440eee57024ffa5f1698956a054/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L355

The logic:
- if resourceVersion is empty, we don't cache
- if requested resourceVesion is old than the cached version (especially for `resourceVersion=0`), serve the cached object.

The cache is in the storage layer using internal versions. I.e. it is not per version. So we webhooks add overhead.

Looking at the watcher code in the apiserver http handler, I am not sure at all that we have a watch cache per version. I see versioning encoders there for the object in the watch events. @liggitt have you looked into that in more detail? My impression is that we cache internal objects and convert on each outgoing stream. I hope I am wrong.

/cc @wojtek-t can you comment our current caching logic?

Reply:
> The cache is in the storage layer using internal versions.

We do not store internal versions, I assume it is one of the external versions in the storage layer?

Reply:
Ok, the "one storage per version" saves us for now. But this important and special: we do not have a hub version, we always convert directly to the target version. Until now this was an arbitrary design choice, i.e. we could have done a NoConversion on decode and the real conversion on encode. With the caching arguments here, we turn this arbitrary choice into a required design. Can you add a remake about this?

Reply:
done